src/utils/generator.js

```diff
@@ -250,7 +251,10 @@ export class Generator {
     invariant(buildNode_ instanceof Function || args.length === 0);
     let id = t.identifier(this.preludeGenerator.nameGenerator.generate("derived"));
     this.preludeGenerator.derivedIds.set(id.name, args);
-    let res = this.realm.createAbstract(types, values, [], id, optionalArgs ? optionalArgs.kind : undefined);
+    let options = {};
+    if (optionalArgs && optionalArgs.kind) options.kind = optionalArgs.kind;
+    let Constructor = Value.isTypeCompatibleWith(types.getType(), ObjectValue) ? AbstractObjectValue : AbstractValue;
+    let res = new Constructor(this.realm, types, values, [], id, options);
```

Comment:
Why not directly via the AbstractValue factory function? I thought we wanted to avoid directly creating AbstractObjectValue/AbstractValue via constructor.

Reply:
There isn't a factory function that is custom made for this situation and since this function is essentially a private helper function for use by AbstractValue, I did not think it worthwhile to create yet another factory function in AbstractValue.